books/RayTracingTheNextWeek.html

```diff
@@ -2057,105 +2057,147 @@
+            } else {
+                // Hunt for the image file in some likely locations.
+                auto filename = std::string(image_filename);
+                if (load(filename)) return;
+                if (load("images/" + filename)) return;
+                if (load("../images/" + filename)) return;
+                if (load("../../images/" + filename)) return;
+                if (load("../../../images/" + filename)) return;
+                if (load("../../../../images/" + filename)) return;
+                if (load("../../../../../images/" + filename)) return;
+                if (load("../../../../../../images/" + filename)) return;
             }
 
-            bytes_per_scanline = bytes_per_pixel * width;
-        }
-
-        ~image_texture() {
-            delete data;
+            std::cerr << "ERROR: Could not load image file '" << image_filename << "'.\n";
+            std::cerr << "STBI: " << stbi_failure_reason() << '\n';
         }
 
-        virtual color value(double u, double v, const vec3& p) const override {
-            // If we have no texture data, then return solid cyan as a debugging aid.
-            if (data == nullptr)
-                return color(0,1,1);
+        ~rtw_image() { STBI_FREE(data); }
 
-            // Clamp input texture coordinates to [0,1] x [1,0]
-            u = interval(0,1).clamp(u);
-            v = 1.0 - interval(0,1).clamp(v);  // Flip V to image coordinates
+        bool load(const std::string filename) {
+            // Loads image data from the given file name. Returns true if the load succeeded.
+            auto components_per_pixel = bytes_per_pixel; // Components per pixel
+            data = stbi_load(filename.c_str(),
+                &image_width, &image_height, &components_per_pixel, bytes_per_pixel);
+            bytes_per_scanline = image_width * bytes_per_pixel;
+            return data != nullptr;
+        }
 
-            auto i = static_cast<int>(u * width);
-            auto j = static_cast<int>(v * height);
+        int width()  const { return (data == nullptr) ? 0 : image_width; }
+        int height() const { return (data == nullptr) ? 0 : image_height; }
 
-            // Clamp integer mapping, since actual coordinates should be less than 1.0
-            if (i >= width)  i = width-1;
-            if (j >= height) j = height-1;
+        const unsigned char* pixel_data(int x, int y) const {
+            // Return the address of the three bytes of the pixel at x,y (or magenta if no data).
+            static unsigned char magenta[] = { 255, 0, 255 };
+            if (!data) return magenta;
 
-            const auto color_scale = 1.0 / 255.0;
-            auto pixel = data + j*bytes_per_scanline + i*bytes_per_pixel;
+            // Clamp pixel coordintes to [0,width-1] x [0,height-1]
+            x = (x < 0) ? 0 : (x < image_width)  ? x : (image_width  - 1);
+            y = (y < 0) ? 0 : (y < image_height) ? y : (image_height - 1);
```

Comment:
we might want to briefly discuss different clamping functions.

Reply:
Consider chunking these calls to make the ternary operator a little more clear
or, just rewriting as if-else statements.

It's pretty clever as is, though. Just not the easiest to parse.

Reply:
does `interval` work here?

Reply:
Keep in mind that this file is really intended as boilerplate code we provide (like the STBI header itself) that the user is not supposed to worry about, particularly since it's so tailored to one particular external. Think of it more as our private extension to STBI to make our lives easier, and to move it off the reader's plate of things to understand.

Also, because of this, I don't have access to the other RTW functions in our codebase (like clamp or interval::clamp). I had stronger wording to the user about this (“don't worry your pretty little head”), but then pulled it because readers who choose to implement this in other languages or with some other image data library will have to wrestle with this somehow anyway.

Still, I'll improve the code above.

Reply:
Ah. Then, yes, I was overthinking it.